pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,229 @@
+func NewMachineDeletionHooksController(
+	operatorClient operatorv1helpers.OperatorClient,
+	machineClient machinev1beta1client.MachineInterface,
+	etcdClient etcdcli.EtcdClient,
+	machineAPIChecker ceohelpers.MachineAPIChecker,
+	masterMachineSelector labels.Selector,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	masterMachineInformer cache.SharedIndexInformer,
+	eventRecorder events.Recorder) factory.Controller {
+	c := &machineDeletionHooksController{
+		machineClient:                         machineClient,
+		etcdClient:                            etcdClient,
+		machineAPIChecker:                     machineAPIChecker,
+		masterMachineSelector:                 masterMachineSelector,
+		masterMachineLister:                   machinelistersv1beta1.NewMachineLister(masterMachineInformer.GetIndexer()),
+		configMapListerForKubeSystemNamespace: kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Lister().ConfigMaps("kube-system"),
+	}
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithInformers(
+			masterMachineInformer,
+			kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Informer(),
+		).
+		ResyncEvery(1*time.Minute). // an arbitrary number, not too high not too low
+		ToController("MachineDeletionHooksController", eventRecorder.WithComponentSuffix("cluster-machine-deletion-hooks-controller"))
+}
+
+func (c *machineDeletionHooksController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	// stop if the machine API is not functional
+	if isFunctional, err := c.machineAPIChecker.IsFunctional(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	var errs []error
+	if err := c.addDeletionHookToMasterMachines(ctx); err != nil {
+		errs = append(errs, err)
+	}
+	// attempt to remove hooks from machines pending deletion and without running etcd member
+	if err := c.attemptToDeleteMachineDeletionHook(ctx, syncCtx.Recorder()); err != nil {
+		errs = append(errs, err)
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+// it will only remove the hook once it has identified that the Machine resource is being deleted and a replacement member has been created.
+func (c *machineDeletionHooksController) attemptToDeleteMachineDeletionHook(ctx context.Context, recorder events.Recorder) error {
+	desiredControlPlaneReplicasCount, err := ceohelpers.ReadDesiredControlPlaneReplicasCount(c.configMapListerForKubeSystemNamespace)
+	if err != nil {
+		return err
+	}
+	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return err
+	}
+	// machines with the deletion hooks are the ones that host etcd members
+	memberMachines := ceohelpers.FilterMachinesWithMachineDeletionHook(masterMachines)
+
+	// if the number of master machines is > controlPlane.replicas based on data from the cache
+	if len(memberMachines) <= desiredControlPlaneReplicasCount {
```

Comment:
Why does the hooks controller care about whether we're over/under the desired control-plane replica count?
I mentioned this in https://github.com/openshift/cluster-etcd-operator/pull/768#discussion_r836926279 but it's the member removal controller that needs to make this consideration when deciding whether it's safe to remove/scale-down a voting member. Plus the member removal controller will also take care of cleaning up learner members pending deletion.

So for the hooks controller, all that would be required is to remove the deletion hook from any machine that isn't mapped to an etcd member (learner or voting).

Reply:
Apologies if I missed this before on my first pass but this just stood out to me when reviewing #768